File: plugins/org.eclipse.tm.tcf/src/org/eclipse/tm/internal/tcf/services/remote/ProcessesProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.io.IOException;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.protocol.JSON;
import org.eclipse.tm.tcf.services.IProcesses;


public class ProcessesProxy implements IProcesses {
    
    private final IChannel channel;
    private final Map<ProcessesListener,IChannel.IEventListener> listeners =
        new HashMap<ProcessesListener,IChannel.IEventListener>();
    
    private class ProcessContext implements IProcesses.ProcessContext {

        private final Map<String,Object> props;
        
        ProcessContext(Map<String,Object> props) {
            this.props = props;
        }

        public String getID() {
            return (String)props.get(PROP_ID);
        }

        public String getParentID() {
            return (String)props.get(PROP_PARENTID);
        }

        public boolean canTerminate() {
            Boolean b = (Boolean)props.get(PROP_CAN_TERMINATE);
            return b != null && b.booleanValue();
        }

        public String getName() {
            return (String)props.get(PROP_NAME);
        }

        public boolean isAttached() {
            Boolean b = (Boolean)props.get(PROP_ATTACHED);
            return b != null && b.booleanValue();
        }

        public IToken attach(final DoneCommand done) {
            return new Command(channel, ProcessesProxy.this,
                    "attach", new Object[]{ getID() }) {
                @Override
                public void done(Exception error, Object[] args) {
                    if (error == null) {
                        assert args.length == 1;
                        error = toError(args[0]);
                    }
                    done.doneCommand(token, error);
                }
            }.token;
        }

        public IToken detach(final DoneCommand done) {
            return new Command(channel, ProcessesProxy.this,
                    "detach", new Object[]{ getID() }) {
                @Override
                public void done(Exception error, Object[] args) {
                    if (error == null) {
                        assert args.length == 1;
                        error = toError(args[0]);
                    }
                    done.doneCommand(token, error);
                }
            }.token;
        }

        public IToken terminate(final DoneCommand done) {
            return new Command(channel, ProcessesProxy.this,
                    "terminate", new Object[]{ getID() }) {
                @Override
                public void done(Exception error, Object[] args) {
                    if (error == null) {
                        assert args.length == 1;
                        error = toError(args[0]);
                    }
                    done.doneCommand(token, error);
                }
            }.token;
        }

        public Map<String, Object> getProperties() {
            return props;
        }

        public String toString() {
            return "[Processes Context " + props.toString() + "]";
        }
    }
    
    public ProcessesProxy(IChannel channel) {
        this.channel = channel;
    }

    public String getName() {
        return NAME;
    }

    public void addListener(final ProcessesListener listener) {
        IChannel.IEventListener l = new IChannel.IEventListener() {

            public void event(String name, byte[] data) {
                try {
                    Object[] args = JSON.parseSequence(data);
                    if (name.equals("exited")) {
                        assert args.length == 2;
                        listener.exited((String)args[0], ((Number)args[1]).intValue());
                    }
                    else {
                        throw new IOException("Processes service: unknown event: " + name);
                    }
                }
                catch (Throwable x) {
                    channel.terminate(x);
                }
            }
        };
        channel.addEventListener(this, l);
        listeners.put(listener, l);
    }

    public void removeListener(ProcessesListener listener) {
        IChannel.IEventListener l = listeners.remove(listener);
        if (l != null) channel.removeEventListener(this, l);
    }

    public IToken getChildren(String parent_context_id, boolean attached_only, final DoneGetChildren done) {
        return new Command(channel, this,
                "getChildren", new Object[]{ parent_context_id, attached_only }) {
            @Override
            public void done(Exception error, Object[] args) {
                String[] ids = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    ids = toStringArray(args[1]);
                }
                done.doneGetChildren(token, error, ids);
            }
        }.token;
    }

    public IToken getContext(String id, final DoneGetContext done) {
        return new Command(channel, this,
                "getContext", new Object[]{ id }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                ProcessContext ctx = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    if (args[1] != null) ctx = new ProcessContext((Map<String, Object>)args[1]);
                }
                done.doneGetContext(token, error, ctx);
            }
        }.token;
    }

    public IToken getEnvironment(final DoneGetEnvironment done) {
        return new Command(channel, this, "getEnvironment", null) {
            @Override
            public void done(Exception error, Object[] args) {
                Map<String,String> env = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    env = toEnvMap(args[1]);
                }
                done.doneGetEnvironment(token, error, env);
            }
        }.token;
    }

    public IToken start(String directory, String file,
            String[] command_line, Map<String,String> environment,
            boolean attach, final DoneStart done) {
        return new Command(channel, this,
                "start", new Object[]{ directory, file, command_line,
                toEnvStringArray(environment), attach }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                ProcessContext ctx = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    if (args[1] != null) ctx = new ProcessContext((Map<String, Object>)args[1]);
                }
                done.doneStart(token, error, ctx);
            }
        }.token;
    }

    public IToken getSignalList(String context_id, final DoneGetSignalList done) {
        return new Command(channel, ProcessesProxy.this,
                "getSignalList", new Object[]{ context_id }) {
            @Override
            public void done(Exception error, Object[] args) {
                Collection<Map<String,Object>> list = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    list = toSignalList(args[1]);
                }
                done.doneGetSignalList(token, error, list);
            }
        }.token;
    }

    public IToken getSignalMask(String context_id, final DoneGetSignalMask done) {
        return new Command(channel, ProcessesProxy.this,
                "getSignalMask", new Object[]{ context_id }) {
            @Override
            public void done(Exception error, Object[] args) {
                int dont_stop = 0;
                int dont_pass = 0;
                int pending = 0;
                if (error == null) {
                    assert args.length == 4;
                    error = toError(args[0]);
                    if (args[1] != null) dont_stop = ((Number)args[1]).intValue();
                    if (args[2] != null) dont_pass = ((Number)args[2]).intValue();
                    if (args[3] != null) pending = ((Number)args[3]).intValue();
                }
                done.doneGetSignalMask(token, error, dont_stop, dont_pass, pending);
            }
        }.token;
    }

    public IToken setSignalMask(String context_id, int dont_stop, int dont_pass, final DoneCommand done) {
        return new Command(channel, ProcessesProxy.this,
                "setSignalMask", new Object[]{ context_id, dont_stop, dont_pass }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    public IToken signal(String context_id, long signal, final DoneCommand done) {
        return new Command(channel, ProcessesProxy.this,
                "signal", new Object[]{ context_id, signal }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneCommand(token, error);
            }
        }.token;
    }

    @SuppressWarnings("unchecked")
    private static String[] toStringArray(Object o) {
        if (o == null) return null;
        Collection<String> c = (Collection<String>)o;
        return (String[])c.toArray(new String[c.size()]);
    }

    private static String[] toEnvStringArray(Map<String,String> m) {
        if (m == null) return new String[0];
        int n = 0;
        String[] arr = new String[m.size()];
        for (String s : m.keySet()) {
            arr[n++] = s + "=" + m.get(s);
        }
        return arr;
    }
    
    @SuppressWarnings("unchecked")
    private static Map<String,String> toEnvMap(Object o) {
        Map<String,String> m = new HashMap<String,String>();
        if (o == null) return m;
        Collection<String> c = (Collection<String>)o;
        for (String s : c) {
            int i = s.indexOf('=');
            if (i >= 0) m.put(s.substring(0, i), s.substring(i + 1));
            else m.put(s, "");
        }
        return m;
    }
    
    @SuppressWarnings("unchecked")
    private static Collection<Map<String,Object>> toSignalList(Object o) {
        if (o == null) return null;
        return (Collection<Map<String,Object>>)o;
    }
}

File: python/src/tcf/services/remote/ProcessesProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

import exceptions
from tcf import channel
from tcf.services import processes
from tcf.channel.Command import Command

class ChannelEventListener(channel.EventListener):
    def __init__(self, service, listener):
        self.service = service
        self.listener = listener
    def event(self, name, data):
        try:
            args = channel.fromJSONSequence(data)
            if name == "exited":
                assert len(args) == 2
                self.listener.exited(args[0], args[1])
            else:
                raise IOError("Processes service: unknown event: " + name);
        except exceptions.Exception as x:
            self.service.channel.terminate(x)

class ProcessesProxy(processes.ProcessesService):
    def __init__(self, channel):
        self.channel = channel
        self.listeners = {}

    def addListener(self, listener):
        l = ChannelEventListener(self, listener)
        self.channel.addEventListener(self, l)
        self.listeners[listener] = l

    def removeListener(self, listener):
        l = self.listeners.get(listener)
        if l:
            del self.listeners[listener]
            self.channel.removeEventListener(self, l)

    def getChildren(self, parent_context_id, attached_only, done):
        service = self
        class GetChildrenCommand(Command):
            def __init__(self):
                super(GetChildrenCommand, self).__init__(service.channel, service, 
                        "getChildren", (parent_context_id, attached_only))
            def done(self, error, args):
                contexts = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    contexts = args[1]
                done.doneGetChildren(self.token, error, contexts)
        return GetChildrenCommand().token

    def getContext(self, context_id, done):
        service = self
        class GetContextCommand(Command):
            def __init__(self):
                super(GetContextCommand, self).__init__(service.channel, service, "getContext", (context_id,))
            def done(self, error, args):
                ctx = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    if args[1]: ctx = ProcessContext(service, args[1])
                done.doneGetContext(self.token, error, ctx)
        return GetContextCommand().token

    def getEnvironment(self, done):
        service = self
        class GetEnvCommand(Command):
            def __init__(self):
                super(GetEnvCommand, self).__init__(service.channel, service, "getEnvironment", None)
            def done(self, error, args):
                env = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    env = _toEnvMap(args[1])
                done.doneGetEnvironment(self.token, error, env)
        return GetEnvCommand().token

    def start(self, directory, file, command_line, environment, attach, done):
        service = self
        env = _toEnvStringArray(environment)
        class StartCommand(Command):
            def __init__(self):
                super(StartCommand, self).__init__(service.channel, service,
                        "start", (directory, file, command_line, env, attach))
            def done(self, error, args):
                ctx = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    if args[1]: ctx = ProcessContext(service, args[1])
                done.doneStart(self.token, error, ctx)
        return StartCommand().token

    def getSignalList(self, context_id, done):
        service = self
        class GetSignalsCommand(Command):
            def __init__(self):
                super(GetSignalsCommand, self).__init__(service.channel, service,
                        "getSignalList", (context_id,))
            def done(self, error, args):
                list = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    list = args[1]
                done.doneGetSignalList(self.token, error, list)
        return GetSignalsCommand().token

    def getSignalMask(self, context_id, done):
        service = self
        class GetSignalMaskCommand(Command):
            def __init__(self):
                super(GetSignalMaskCommand, self).__init__(service.channel, service,
                        "getSignalMask", (context_id,))
            def done(self, error, args):
                dont_stop = 0
                dont_pass = 0
                pending = 0
                if not error:
                    assert len(args) == 4
                    error = self.toError(args[0])
                    dont_stop, dont_pass, pending = args[1:3]
                done.doneGetSignalMask(self.token, error, dont_stop, dont_pass, pending)
        return GetSignalMaskCommand().token

    def setSignalMask(self, context_id, dont_stop, dont_pass, done):
        service = self
        class SetSignalMaskCommand(Command):
            def __init__(self):
                super(SetSignalMaskCommand, self).__init__(service.channel, service,
                        "setSignalMask", (context_id, dont_stop, dont_pass))
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneCommand(self.token, error)
        return SetSignalMaskCommand().token

    def signal(self, context_id, signal, done):
        service = self
        class SignalCommand(Command):
            def __init__(self):
                super(SignalCommand, self).__init__(service.channel, service,
                        "signal", (context_id, signal))
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneCommand(self.token, error)
        return SignalCommand().token

def _toEnvStringArray(map):
    arr = []
    if not map: return arr
    for name, value in map.items():
        arr.append("%s=%s" % (name, value))
    return arr

def _toEnvMap(arr):
    map = {}
    if not arr: return map
    for str in arr:
        i = str.find('=')
        if i >= 0: map[str[:i]] = str[i + 1:]
        else: map[str] = ""
    return map

class ProcessContext(processes.ProcessContext):
    def __init__(self, service, props):
        super(ProcessContext, self).__init__(props)
        self.service = service

    def attach(self, done):
        return self._command("attach", done)

    def detach(self, done):
        return self._command("detach", done)

    def terminate(self, done):
        return self._command("terminate", done)

    def _command(self, command, done):
        service = self.service
        id = self.getID()
        class _Command(Command):
            def __init__(self):
                super(_Command, self).__init__(service.channel, service,
                        command, (id,))
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneCommand(self.token, error)
        return _Command().token
